src/stm32f411/clock.cpp

```diff
@@ -0,0 +1,425 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <libhal-arm-mcu/stm32f411/clock.hpp>
+
+#include <cstdint>
+
+#include <libhal-util/bit.hpp>
+#include <libhal-util/enum.hpp>
+#include <libhal/error.hpp>
+#include <libhal/units.hpp>
+
+#include "flash_reg.hpp"
+#include "libhal-arm-mcu/stm32f411/constants.hpp"
+#include "power.hpp"
+#include "rcc_reg.hpp"
+#include "stm32f411/power.hpp"
+namespace hal::stm32f411 {
+
+namespace {
+hal::hertz rtc_clock_rate = 0.0_Hz;      // defaults to "no clock"
+hal::hertz pll_clock_rate = 0.0_Hz;      // undefined until pll is enabled
+hal::hertz usb_otg_clock_rate = 0.0_Hz;  // undefined until enabled
+hal::hertz ahb_clock_rate = internal_high_speed_oscillator;
+hal::hertz apb1_clock_rate = internal_high_speed_oscillator;
+hal::hertz apb2_clock_rate = internal_high_speed_oscillator;
+hal::hertz timer_apb1_clock_rate = internal_high_speed_oscillator;
+hal::hertz timer_apb2_clock_rate = internal_high_speed_oscillator;
+}  // namespace
+
+void pll_valid(clock_tree p_clock_tree)
+{
+  /// Check if the VCO input can be set from 1_MHz to 2_MHz
+  if (p_clock_tree.pll.source == pll_source::high_speed_external) {
+    if (p_clock_tree.high_speed_external < 2.0_MHz ||
+        p_clock_tree.high_speed_external > 63 * 2.0_MHz) {
+      hal::safe_throw(hal::argument_out_of_domain(nullptr));
+    }
+  }
+
+  /// Check the Output Frequency is within bound
+  if (p_clock_tree.pll.output > (100.0_MHz) ||
+      p_clock_tree.pll.output < (100.0_MHz / 8)) {
+    hal::safe_throw(hal::argument_out_of_domain(nullptr));
+  }
+}
+
+// TODO (#62): Add I2S to the clock configs
+// TODO (#66): Add 48_MHz to the clock configs
+void configure_pll(clock_tree p_clock_tree)
+{
+  if (p_clock_tree.pll.enable) {
+    auto vco_in = 2.0_MHz;
+    uint8_t input_division_factor = 2U;
+    pll_valid(p_clock_tree);
+
+    if (p_clock_tree.pll.source == pll_source::high_speed_external) {
+      if (p_clock_tree.high_speed_external < 4.0_MHz) {
+        vco_in = p_clock_tree.high_speed_external / 2;
+        bit_modify(rcc->pllconfig)
+          .insert(rcc_pllcnfg::input_division_factor, input_division_factor);
+      } else {
+        float division_factor = p_clock_tree.high_speed_external / 2.0_MHz;
+        if (division_factor - static_cast<uint8_t>(division_factor) > 0) {
+          input_division_factor = static_cast<uint8_t>(division_factor) + 1;
+        } else {
+          input_division_factor = static_cast<uint8_t>(division_factor);
+        }
+        vco_in = p_clock_tree.high_speed_external / division_factor;
+      }
+    } else {
+      input_division_factor = 8;
+    }
+    bit_modify(rcc->pllconfig)
+      .insert(rcc_pllcnfg::input_division_factor, input_division_factor);
+
+    uint8_t output_division_factor = 0;
+    while (p_clock_tree.pll.output * (2 * (output_division_factor + 1)) <
+           100.0_MHz) {
+      output_division_factor++;
+    }
```

Comment:
Why do we need to multiply by 2? Can these be constants? Or can you link the location of this equation in the datasheet with a comment.